### PR TITLE
Session-based example using dressipi dataset and XLNet architecture

### PR DESCRIPTION
**Pull request by rnyak**
This PR is addressing session-based example creation task in https://github.com/NVIDIA-Merlin/models/issues/734.

Currently 
- This PR modifies the existing [session-based example](https://github.com/NVIDIA-Merlin/models/blob/main/examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb)  notebook. We could remove  the custom biLSTM class and we can directly use [tf.keras.layers.Bidirectional](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Bidirectional) instead.
-  we can run XLNET architecture however weight-tying and loader does not work properly (see blockers below) unless we go with workaround solutions.

Blockers:
we need these two tickets to be solved 
- https://github.com/NVIDIA-Merlin/models/issues/848
- https://github.com/NVIDIA-Merlin/models/issues/844

### CODE REVIEW

File: examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb [hunk 1]
```diff
@@ -99,14 +99,21 @@
      "name": "stderr",
```
Comment (marcromeyn) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:99]: <p>Is there anything blocking us for only using the V2-API in this notebook?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='63'/>

Comment (rnyak) [reply examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:99]: I will update to V2 api :) and we'll see.

File: examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb [hunk 2]
```diff
@@ -70,7 +70,7 @@
     "\n",
```
Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>Why do you remove the tags.Target from purchase_id_first? that is the only difference between target_feature and groupby_features['purchase_id_first']</p><p>If you want this change, can we delete following code :</p><pre class="ql-syntax" spellcheck="false">target_feature = (
&nbsp; &nbsp; groupby_features['purchase_id_first'] &gt;&gt;&nbsp;
&nbsp; &nbsp; AddMetadata(tags=[Tags.TARGET])
)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>The MRR drops from 0.145 to 0.0922 -&gt; this is 37% lower than before.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='61'/>

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>Why did we drop the header "Training a Bi-LSTM with Merlin Models" Before the section? I think we should add something to explain, that we have a new section</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>I think this text is not uptodate with the new API, anymore. Is there a MultiClassClassificationTask ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>Yes, the MRR improved :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='83'/>

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>Can we add a architecture diagram, fitting the style above?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='85'/>

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>Can we not select by Tag sequence? instead of schema_model.select_by_name(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;['item_id_list_seq', 'f_47_list_seq', 'f_68_list_seq']</p><p> </p><p>Similar, do we have a tag for schema_model.select_by_name(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;['item_id_last']</p><p>&nbsp;&nbsp;&nbsp;&nbsp;) ?</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='88'/>

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>can we make the hidden dimension a hyperparameter variable and use it? e.g. hiddem_dim or d_dim ? so it is easy to stop which values have to be the same?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='94'/>

**Comment by bschifferer** [comment]
Added comments - overall looks pretty good.

File: examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb (hunk 2 above)
Comment (rnyak) [reply examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>thanks. yes, I think we can remove the code above you pointed put. the reason I removed it because of the transform in the dataloader. I was told  <code>ToTarget</code> method transform column to target to tag is not needed. </p>

Comment (rnyak) [reply examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>I think MRR before is <code>0.1185</code> not really <code>0.145</code> . Not sure why it printed it out different in the current example? do you know the reason?  </p>

Comment (rnyak) [reply examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: In the new nb I only train for 3 epochs, whereas the old one trained for 10 epochs. I am testing with 10 epochs now with the same hyper-params. let's see.

Comment (bschifferer) [inline examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>mm.MLPBlock(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[128,256],</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;activation='relu',</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;no_activation_last_layer=True,</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;dropout=DROPOUT,</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;)</p><p> </p><p>should the 256 not be d_model as well?</p><p> </p><p>My suggestion was that it is easier to see that the MLPBlock last hidden dimension and Transformer dimension have to align</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/849/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='96'/>

**Review by bschifferer — APPROVED** [review]
Looks great! I added one small comment to use the dmodel parameter for the first MLP block as well

File: examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb (hunk 2 above)
Comment (rnyak) [reply examples/usecases/ecommerce-session-based-next-item-prediction-for-fashion.ipynb:70]: <p>ok. I can modify that.</p>

**Comment by rnyak** [comment]
rerun tests

File: tests/unit/tf/examples/test_usecase_ecommerce_session_based.py [hunk 3]
```diff
@@ -59,3 +59,16 @@ def test_usecase_ecommerce_session_based(tb):
             "regularization_loss",
         ]
     )
+    metrics_transformer = tb.ref("metrics_transformer")
+    assert set(metrics_transformer.keys()) == set(
```
Comment (oliverholworthy) [inline tests/unit/tf/examples/test_usecase_ecommerce_session_based.py:63]: Can you add `pytest.importorskip("transformers")` to the top of this file to reflect the dependency on transformers now

Comment (rnyak) [reply tests/unit/tf/examples/test_usecase_ecommerce_session_based.py:63]: yes, will do. thanks!
